### DISCUSSION TITLE
Storing user state in the player

### DISCUSSION THREAD

**Discussion by SamithDB** [Lumieducation/H5P-Nodejs-library#1589] [Q&A] [answered]
Hi, I'm developing h5p app with H5P-Nodejs-library and I need to save answer data after user try any content and I need to view those data in another h5p player. Any tips to achieve this? I saw "storing user state in the player" is not finished yet according to Project Status page. 
Thank you.

**Comment by sr258** [comment, marked as answer]
This feature is indeed not implemented yet. It shouldn't be terribly complex, but we haven't found time for it yet. We might get funding for it at the end of the year, bit that's still uncertain.

You can always work on the feature yourself if it's urgent. We can provide some tips.

**Comment by emmanuelpineda** [comment]
Hello, could you tell me if you plan to develop this feature, if not, could you provide me with some tips to develop it myself please?
For the project I am developing it is necessary for the user to be able to go back and see what It has previously entered.

Thank you in advance.

**Comment by sr258** [comment]
The feature is now part of 9.2.0. See the docs here: https://docs.lumi.education/advanced-usage/user-content-state https://docs.lumi.education/advanced-usage/completion-tracking

**Reply by emmanuelpineda** [reply]
I will check the documentation, Thank you very much!

**Comment by emmanuelpineda** [comment]
I have enabled the option to save user content state, it works fine. I see that by default the state is saved to the database with the option preload = true, however, how can I set this option to false to load the state with a second AJAX request from the browser?

I tried to set the preload value from the options of the H5P.setUserData function defined in the h5p.js file but it doesn't work

![image](https://user-images.githubusercontent.com/45979790/176609240-6cfbe7f7-983d-42fb-9f8a-95e23059d683.png)

I am trying to use this option to change the functionality a bit because according to the project requirements it is necessary that the admin can reuse the H5P content in different sections, however, when the user responds the content and the state is saved , the state will be restored in all sections where that content is used.

Could you advise me please?

Thank you in advance.

**Reply by sr258** [reply]
The `preloaded` flag is set by the H5P client and I believe that the player usually sets it to `true`. I'm not sure if the player even works if the state isn't preloaded. The only case in which a state isn't preloaded at the moment, is the editor.

Even if you re-use content in several places, isn't the state still restored with `preloaded` set to true?

**Reply by emmanuelpineda** [reply]
Yes, the state is restored, but I would like to keep a separate state for each section even though the H5P content is the same. For this, I was planning to modify the contentUserDataAjax function in the h5p.js file a bit to concatenate an extra parameter to the id of the content. However, the state is not restored when preload = false

**Reply by sr258** [reply]
You can implement the POST route of the content user state yourself and always set preloaded to false there. You‘ll have to try if the player loads the state with an Ajax call then, however.

**Reply by sr258** [reply]
I've dug into this and setting the `preloaded` option to false all the time works for the server-side. The h5p.js file is "buggy" however, in the way that the init function doesn't expect the getUserState function to be async. It simply continues with initialization without waiting for the user state GET request to be finished. I've created a simple "hacky" fix that you can use as a base: https://github.com/h5p/h5p-php-library/compare/master...sr258:h5p-php-library:fix/async-state

**Reply by emmanuelpineda** [reply]
Thank you very much, I will try it. Should I wait for this change to be applied to the main branch?

I've been testing setting the async = false parameter when calling the contentUserDataAjax function, however I don't know if this is correct or has negative implications. This way seems to work too.

![image](https://user-images.githubusercontent.com/45979790/176760300-e0e7cc8c-7d9f-4ffd-bf35-7e0aed736241.png)

**Reply by sr258** [reply]
The h5p.js file isn’t included in our repo but it’s downloaded from the h5p-php-library repo. So you’ll have to patch it yourself. I don’t think there are negative effects of making the one call sync except that the app might load a (small) but slower.
